pkg/git/strategy/proxy/strategy_proxy_test.go

```diff
@@ -0,0 +1,282 @@
+					userAgent := ctx.Req.Header.Get("User-Agent")
+					if strings.Contains(host, "example.com") && strings.Contains(userAgent, "libgit2") {
+						*proxyGotRequest = true
+						return goproxy.OkConnect, u.Host
+					}
+					//reject if it isnt our request
+					return goproxy.RejectConnect, host
+				}
+				proxy.OnRequest().HandleConnect(proxyHandler)
+
+				return &git.AuthOptions{
+						Transport: git.HTTPS,
+						Username:  username,
+						Password:  password,
+						CAFile:    exampleCA,
+					}, func() {
+						os.RemoveAll(gitServer.Root())
+						gitServer.StopHTTP()
+					}
+			},
+			shortTimeout:  false,
+			wantUsedProxy: true,
+			wantError:     false,
+		},
+		{
+			name:    "gogit_HTTP_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "http://example.com/bar/test-reponame",
+			branch:  "main",
+			setupGitProxy: func(g *WithT, proxy *goproxy.ProxyHttpServer, proxyGotRequest *bool) (*git.AuthOptions, cleanupFunc) {
+				// Create the git server.
+				gitServer, err := gittestserver.NewTempGitServer()
+				g.Expect(err).ToNot(HaveOccurred())
+
+				username := "test-user"
+				password := "test-password"
+				gitServer.Auth(username, password)
+				gitServer.KeyDir(gitServer.Root())
+
+				g.Expect(gitServer.StartHTTP()).ToNot(HaveOccurred())
+
+				// Initialize a git repo.
+				err = gitServer.InitRepo("../testdata/repo1", "main", "bar/test-reponame")
+				g.Expect(err).ToNot(HaveOccurred())
+
+				u, err := url.Parse(gitServer.HTTPAddress())
+				g.Expect(err).ToNot(HaveOccurred())
+
+				var proxyHandler goproxy.FuncReqHandler = func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
+					userAgent := req.Header.Get("User-Agent")
+					if strings.Contains(req.Host, "example.com") && strings.Contains(userAgent, "git") {
+						*proxyGotRequest = true
+						req.Host = u.Host
+						req.URL.Host = req.Host
+						return req, nil
+					}
+					//reject if it isnt our request
+					return req, goproxy.NewResponse(req, goproxy.ContentTypeText, http.StatusForbidden, "")
+				}
+				proxy.OnRequest().Do(proxyHandler)
+
+				return &git.AuthOptions{
+						Transport: git.HTTP,
+						Username:  username,
+						Password:  password,
+					}, func() {
+						os.RemoveAll(gitServer.Root())
+						gitServer.StopHTTP()
+					}
+			},
+			shortTimeout:  false,
+			wantUsedProxy: true,
+			wantError:     false,
+		},
+		{
+			name:    "gogit_HTTPS_PROXY",
+			gitImpl: gogit.Implementation,
+			url:     "https://github.com/git-fixtures/basic",
+			branch:  "master",
+			setupGitProxy: func(g *WithT, proxy *goproxy.ProxyHttpServer, proxyGotRequest *bool) (*git.AuthOptions, cleanupFunc) {
+				var proxyHandler goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
+					userAgent := ctx.Req.Header.Get("User-Agent")
+					if strings.Contains(host, "github.com") && strings.Contains(userAgent, "Go-http-client") {
```

Comment:
Maybe you can combine this with the comment in https://github.com/fluxcd/source-controller/pull/524/files#diff-5f975bcd01803cd41840c25795d6ffee1c6a54591ddb10e7a252edae067e9ea1R187-R188 . They are the same, just moving the comment above and adding a line about the user-agent should be enough.

Reply:
I'm not sure this has to do with the CaBundle/custom Transport issue
Rather, I believe it's because `ctx.Req` refers to the `CONNECT` request made to the proxy, and not the HTTPS request to the Git server initiated by go-git (called `req` in the HTTP handler)
Accessing the underlying HTTPS Git request would require a MITM setup

Transports don't carry headers, only Requests do, and go-git applies the `User-Agent` header to all of them [here](https://github.com/go-git/go-git/blob/a5bbcd278ab1467bf9a8dfa615649b6c4d699119/plumbing/transport/http/common.go#L21)
Since Go's net/http is the one making a new `CONNECT` request on behalf of go-git, the headers are not copied (not to expose sensitive data to the proxy server, and because that connection is often through plain HTTP)

IMHO, it would be a good idea to add a separate comment to make clearer the meaning of `ctx.Req`
The user agent check could also potentially be removed altogether since it's not very accurate (any request using Go's net/http would trigger a false positive) and can be a bit misleading when reading the code, but then an explanation would have to be added as to why it is not present

Reply:
You're right. Just checked the method and it is a `CONNECT` request. Thanks again for the correction :)